miden-crypto/src/merkle/smt/historical/mod.rs

```diff
@@ -0,0 +1,500 @@
+    inner: RwLockReadGuard<'a, InnerState>,
+    /// The reversions needed to reach this historical state from `latest`.
+    ///
+    /// `latest + [0, 1, 2, 3, 4, 5... MAX_HISTORY] -> state at[n-1] steps into the past`
+    /// where `n` is implicitly encoded in the `reversions.len()`.
+    reversions: Vec<Arc<HistoricalReversion>>,
+}
+
+impl<'a> HistoricalView<'a> {
+    /// An iterator for the reversion stack
+    ///
+    /// Flattens all mutations from all historical states, to be applied in order.
+    /// The order is from newest to oldest.
+    fn reversion_iter(
+        &self,
+    ) -> impl DoubleEndedIterator<Item = &MutationSet<SMT_DEPTH, Word, Word>> + '_ {
+        self.reversions.iter().map(|state| state.mutations.as_ref())
+    }
+
+    /// Root of the historical view
+    pub fn root(&self) -> Word {
+        // If we have reversions, the last one contains the historical root we want
+        self.reversions
+            .last()
+            .map(|state| state.root)
+            .unwrap_or_else(|| self.inner.latest.root())
+    }
+
+    /// Wrapper.
+    pub(crate) fn key_to_leaf_index(key: &Word) -> LeafIndex<SMT_DEPTH> {
+        <Smt as SparseMerkleTree<SMT_DEPTH>>::key_to_leaf_index(key)
+    }
+
+    /// Lookup an inner node directly from the reversions
+    fn lookup_inner_node_from_reversions(&self, node_index: NodeIndex) -> Option<InnerNode> {
+        // Check reversions in reverse order _oldest to newest_ and use the OLDEST reversion that
+        // mentions this node.
+        for reversion in self.reversion_iter().rev() {
+            if let Some(mutation) = reversion.node_mutations().get(&node_index) {
+                match mutation {
+                    NodeMutation::Removal => {
+                        // Node was removed - return empty subtree root
+                        let depth = node_index.depth();
+                        return Some(EmptySubtreeRoots::get_inner_node(SMT_DEPTH, depth));
+                    },
+                    NodeMutation::Addition(inner_node) => {
+                        // Node was added/modified - return the stored value
+                        return Some(inner_node.clone());
+                    },
+                }
+            }
+        }
+        // No mutation found in reversions
+        None
+    }
+
+    fn get_inner_node(&self, index: NodeIndex) -> InnerNode {
+        // Check reversions for mutations
+        if let Some(inner_node) = self.lookup_inner_node_from_reversions(index) {
+            return inner_node;
+        }
+
+        // Fall back to latest tree
+        self.inner.latest.get_inner_node(index)
+    }
+
+    /// Get the hash of a node at an arbitrary index, including the root or leaf hashes.
+    fn get_node_hash(&self, index: NodeIndex) -> Word {
+        if index.is_root() {
+            return self.root();
+        }
+
+        let mut found_mutation = None;
```

Comment:
Curious: Could hash-based early termination eliminate the O(h) reversion scan here? Pre-compute a HashMap cache in apply_mutations() and check it before iterating through all 33 reversions. Might reduce lookups from O(h) to O(1) for cached nodes.